Added support to register nested classes and reflection via classNames

lower-case opened the pull request:
Resolves #9109 

Added support for registering for Reflection  via classNames.
Now users can register private classes which were inaccessible otherwise.

lower-case committed:
Added support to register nested classes and reflection via classNames

geoand commented:
Thanks for this!

I'll only add a couple small comments for now since I'm not on a phone and it's hard to review on a small screen.

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:
```diff
@@ -30,18 +33,48 @@
+            if (namesValue != null) {
+                String[] names = namesValue.asStringArray();
+                for (String name : names) {
+                    if (isClass(name)) {
```
Why the check here?

lower-case replied:
Wasn't sure how the handling is done afterwards, hence added a validation check. I've removed the validation as per your suggestion.

geoand replied:
:+1

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:
```diff
@@ -30,18 +33,48 @@
+
+    private boolean isClass(String className) {
+        try {
+            Class.forName(className, false, getClass().getClassLoader());
```
If you absolutely do need to load a class in a Quarkus build step, you should use the Thread Context Classloader.

geoand commented on core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:
```diff
@@ -30,4 +30,11 @@
+     * This allows for classes to be registered via class names. This was introduced to add the facility to
+     * register private classes for Reflection.
+     */
+    String[] names() default {};
```
Perhaps calling it `classNames` would be better?

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:
```diff
@@ -30,18 +33,48 @@
+            Class.forName(className, false, getClass().getClassLoader());
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
```
If we are going to load classes (I assume to make things safer for users, we should at least be logging when the load failed.

But as I said above, I don't really see much point in this type of validation)

lower-case replied:
Removed the Validation, This won't be required now.

lower-case commented:
Hey @geoand, Thanks for your super quick reply and valuable feedback 💯 

I've made some changes as per your suggestions.
Also, I've made some extra changes like removing the Exception from method definition and removing this line as the variable wasn't used anywhere
`BuildProducer<RuntimeInitializedClassBuildItem> runtimeInit;`
Please advise if I should revert these to original and review the new changes :)

geoand commented on core/runtime/src/main/java/io/quarkus/runtime/annotations/RegisterForReflection.java:
```diff
@@ -30,4 +30,10 @@
     Class<?>[] targets() default {};
+
+    /**
+     * This allows for classes to be registered via class names. This was introduced to add the facility to
```
```suggestion
     * This allows for classes to be registered for reflection via class names. This is useful when it's necessary to
```

geoand commented:
You are welcome!

I think this looks good now, feel free to take it out of draft

rsvoboda reviewed (commented):
+1 for comments from @geoand, have nothing else to add at this stage. Please make it regular PR so all the checks CI jobs are executed.

`isClass` in the previous version wasn't working for `org.kohsuke.github.GHRepository.GHRepoPermission` case I was struggling with as it is private inner class - https://github.com/hub4j/github-api/blob/master/src/main/java/org/kohsuke/github/GHRepository.java#L184

rsvoboda commented:
@geoand I have hit another issue with `github-api` library in native mode - https://stackoverflow.com/questions/61693823/quarkus-native-and-org-kohsukegithub-api-unresolved-method-during-parsing ... if you have time to take a look ...

geoand commented:
@rsvoboda sure thing, I'll take a look later on

lower-case commented:
Awesome!! Thanks for the help guys. As suggested by @geoand, I've moved the PR out of Draft. Please advise on the next steps. Do I have to work on tests and If yes Can you please provide a reference  to which I can get an idea of testing Framework. 
Thanks in advance!!

geoand commented:
I think the best test would be to actually use the new method of the annotation in the main integration test.

You can add a new JAX-RS endpoint, that tries to read a method of some new class with reflection and just return the name for example.
That class would be registered for reflection with using the class name.

Does that make sense?

lower-case commented:
Hey @geoand , I've added some test cases as you suggested. 
Also, I added an additional feature to include nested classes for registration because this is simpler and less error prone than usage of registration via classNames. Please review the changes and provide your suggestions.

geoand reviewed (approved):
Great stuff, thanks a lot!

geoand commented:
You'll need to sort the import of `RegisterForReflectionITCase`.

Just run `mvn process-resources -f integration-tests/main`, amend the commit with the changes and force-push.

geoand reviewed (changes requested):
Actually, the tests won't really test what you think they do.

The code that tries to load the class should be inside the application itself. 
You can use a JAX-RS endpoint to return the class name for example and assert on that.
Otherwise the native tests you have added don't make sense.

geoand commented on integration-tests/main/src/main/java/io/quarkus/it/runtime/ReflectionResource.java:
```diff
@@ -0,0 +1,24 @@
+import javax.ws.rs.QueryParam;
+
+/**
+ * User : lovekeshgarg
```
Please remove these comments

lower-case commented:
Hey @geoand , I really appreciate your thorough review !! I've pushed my changes and also I was able to tests the fix locally via Integration Suite. Please let me know If any more changes are required.

geoand reviewed (approved):
Great stuff, thanks a lot!

geoand commented on core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java:
```diff
@@ -27,22 +26,62 @@
+                registerClass(classLoader, clazz.getName(), methods, fields, false);
+            }
+        } catch (ClassNotFoundException e) {
+            log.errorf(e, "Failed to load Class %s", className);
```
Actually, this should probably be a warning, not an error

lower-case commented:
Thanks @geoand  for your help in getting this merged.
It was my first contribution to an Open Source Project. 🙏 

I browsed through the issues and found these. Have pushed a draft PR for one of them. Kindly provide your suggestions on these:
https://github.com/quarkusio/quarkus/issues/9121
https://github.com/quarkusio/quarkus/issues/9122

geoand commented:
Glad you enjoyed the experience.

I'll let the subject matter expert comment on your other Pull Request
